### ISSUE TITLE
`polkadot-archive`: will stuck with log `Doesn't contain metadata`

### ISSUE THREAD

**Issue by koushiro** [paritytech/substrate-archive#161]
I encountered the following log when I run the `polkadot-archive` (eac41861b674cf90b1f7ce8de057b64928ea7851) every time  from scratch. It can be seen from the log that `batch_block_handler` has fallen into an infinite loop. It seems to work normally again when I restart `polkadot-archive`.

## Command
```
./polkadot-archive -c polkadot.toml --chain=polkadot -vvv 
```

## Log
```
[01:34] INFO Doesn't contain metadata
[01:34] INFO Doesn't contain metadata
...
[10:15] INFO Doesn't contain metadata
[10:15] INFO Doesn't contain metadata
```

## batch_block_handler

```rust
	// Returns true if all versions are in database
	// false if versions are missing
	async fn db_contains_metadata(blocks: &[Block<B>], conn: &mut DbConn) -> Result<bool> {
		let specs: hashbrown::HashSet<u32> = blocks.iter().map(|b| b.spec).collect();
		let versions: hashbrown::HashSet<u32> = queries::get_versions(conn).await?.into_iter().collect();
		Ok(specs.is_subset(&versions))
	}

	async fn batch_block_handler(&self, blks: BatchBlock<B>) -> Result<()>
	where
		NumberFor<B>: Into<u32>,
	{
		let mut conn = self.db.conn().await?;
		while !Self::db_contains_metadata(blks.inner(), &mut conn).await? {
			log::info!("Doesn't contain metadata");
			smol::Timer::after(Duration::from_millis(50)).await;
		}
		std::mem::drop(conn);
		self.db.insert(blks).await?;
		Ok(())
	}
```

## PostgreSQL
```sql

polkadot-archive=> select spec,min(block_num),max(block_num),count(*) from blocks 
polkadot-archive=> where spec in (select version from metadata)
polkadot-archive=> group by spec;
 spec |  min   |  max   | count  
------+--------+--------+--------
    0 |      1 |  29230 |  29230
    1 |  29231 | 188835 | 159605
    5 | 188836 | 199404 |  10569
    6 | 199405 | 200000 |    596
(4 rows)

polkadot-archive=> select count(*) from blocks;
 count  
--------
 200000
(1 row)
```

**Comment by koushiro** [comment]
I have added some logs. The above problem seems to be caused by calling `pool.acquire()` to get a connection, which causes a failure to insert data into the database.

```
[11:25] ERROR [insert] Acquire database pool error: pool timed out while waiting for an open connection, pool
 state: Pool { size: 13, num_idle: 0, is_closed: false, options: PoolOptions { max_connections: 28, min_conne
ctions: 4, connect_timeout: 30s, max_lifetime: Some(1800s), idle_timeout: Some(3.6s), test_before_acquire: tr
ue } }
[11:25] ERROR Insert metadata into database error, sqlx error: pool timed out while waiting for an open conne
ction
[11:25] ERROR [conn] Acquire database pool error: pool timed out while waiting for an open connection, pool s
tate: Pool { size: 12, num_idle: 0, is_closed: false, options: PoolOptions { max_connections: 28, min_connect
ions: 4, connect_timeout: 30s, max_lifetime: Some(1800s), idle_timeout: Some(3.6s), test_before_acquire: true
  } }
[11:25] ERROR Insert batch storage error, sqlx error: pool timed out while waiting for an open connection
```

**Comment by insipx** [comment]
This is definitely interesting (and I haven't gotten this myself before) -- especially since the logs say that the pool size was 12 which isn't even close to the 28 max connections specified. At first glance this seems like something in the codebase that's starving futures of acquiring a database connection to the pool? I would have to look into it more to see if that's true, but that'd be my first guess.

**Comment by koushiro** [comment]
it had been fixed in #186.

The reason is that `await` is missing in the `batch_block_handler` of `actors/workers/metadata.rs`.
